styles/Canonical/016-No-inline-comments.yml

```
@@ -4,9 +4,7 @@ scope: raw
 nonword: true
 level: warning
 tokens:
-  - ^```.*(\n(?s:.)*#.*)+.*\n```
-  - ^```.*(\n(?s:.)*//.*)+.*\n```
+  - ^```.*\n([^\n][^\n#]*\n)*([^\n#]+#.*\n)+([^\n#]*\n)*```
```

`^```.*` will match not only code blocks, but all directives in MyST.
Which means (I think) that this will fire on any `#`, for example a link to a page anchor.

---

That's likely true, but this is a fix of an issue with this rule that causes vale to hang when processing MD files - and has nothing to do with incorrectly catching MyST syntax, or the coverage of this rule.

---

Essentially, the previous rule declaration was causing MD files of any significant length to hang in processing. For example, if you use the main branch of the starter pack and use the Vale entrypoint, or test against the MyST syntax guide.

I'm not entirely sure why, I assume it's to do with the nesting, or the non-capturing group. I just implemented the same logic in a different way - trying not to nest groups as much. Testing validates the change.

---

Yes, that part I understand. I meant what you want the expression to do - I see three groups, the first and last optional, with a (seemingly) random distribution of "anything except line break" and "anything except line break or #".

But maybe I'm just the wrong person to review this. ;)

---

Your perspective is greatly appreciated, and you often see things that others don't with these - but the purpose of this PR is not to address false positives for this rule. It is to address the larger bug which causes the vale action to hang - which prevent vale from running on realistic MD targets.

The issue you've brought up here should be addressed, but it should be addressed with its own issue and PR.

The core of this PR is 'do the changed lines implement the same logic?' and 'does the change solve the issue of vale hanging?'.

1. `^```.*\n` Identifies the standard Markdown code block opening syntax. As you have noted, it will also capture MyST directives, but it is identical to the rule that has already been accepted.
2. `([^\n][^\n#]*\n)*` Identifies the potential for a line that may start with a comment `#` (hence `[^\n]`), but it should not capture a line that has a comment inline after some content (hence `[^\n#]*`). This line allows for normal content in the code block before the section we are looking for in this issue.
3. `([^\n#]+#.*\n)+` Identifies what the rule is looking for, an inline comment.
4. `([^\n#]*\n)*` Allows any amount of content after the line we are looking for (which actually now that I'm examining it - it should be `([^\n]*\n)*`.
5. Ends the code block.

This is a reimplementation of the existing logic, which does not address edge cases in this rule. You are correct though, there are significant improvements to make to this rule, but this PR is for a bugfix.

---

```suggestion
  - ^```.*\n(([^\n`][^\n#]*\n|\n))*([^\n#]+#.*\n)+([^\n`]*\n)*```
```

Avoids looking at more than a single instance of a code block. Lines can't contain code block delimiters, also allows for a blank line in the code before the identified line with the inline comment.

---

> * `` ^```.*\n `` Identifies the standard Markdown code block opening syntax. As you have noted, it will also capture MyST directives, but it is identical to the rule that has already been accepted.
> * `([^\n][^\n#]*\n)*` Identifies the potential for a line that may start with a comment `#` (hence `[^\n]`), but it should not capture a line that has a comment inline after some content (hence `[^\n#]*`). This line allows for normal content in the code block before the section we are looking for in this issue.
> * `([^\n#]+#.*\n)+` Identifies what the rule is looking for, an inline comment.
> * `([^\n#]*\n)*` Allows any amount of content after the line we are looking for (which actually now that I'm examining it - it should be `` ([^\n`]*\n)* ``.
> * Ends the code block.

Thanks - this is the explanation I was looking for. :)

How about code blocks that have a full comment line and an inline comment, but with the inline comment first?
Sounds like your regex only catches cases where the full comment line comes first.

````
```
1 + 1 # simple math
# And now to something more difficult:
1 * 1
```
````

---

I believe that as long as there's an inline comment, it will be caputred.

Group one accepts a comment, but not an inline one because a normal comment is acceptable (though our style guide disagrees - again, I'm just reimplementing the existing rule). This group allows 0 of these occurances as well, so the inline comment can be found first.

The actual important group is the second, the one that identifies the first line in your example. It is required, though honestly it doesn't need to be repeated (as capturing these groups is largely irrelevant anyway as we're not able to make sensible suggestions around the captured groups).

The third group then allows anything, excluding the end of code delimiter, because the target has already been found. This includes the possibility of a comment, or an inline comment.

I should add that my reimplementation is based on the test cases that were originally included with this rule.

Moving forward, I think the rules need an additional review before full release. I believe the intent is to do some practical testing on sets of documentation (hence wanting the prpcessing bug to be fixed as a priority), but that practical testing will likely capture false positives rather than uncaptured issues.

---

I've opened #55 to address edge cases and uncaptured cases.

---

> The third group then allows anything, excluding the end of code delimiter, because the target has already been found. This includes the possibility of a comment, or an inline comment.

Yeah okay, with your update it does. I was looking at the explanation/initial suggestion.